### PR TITLE
Support monorepos with independent versions

### PR DESCRIPTION
**Pull request by rickycodes**
This changes functionality to support MetaMask monorepos with an independent versioning strategy.

closes: #45

### CODE REVIEW

File: .github/workflows/build-lint-test.yml [hunk 1]
```diff
@@ -16,9 +16,48 @@ jobs:
...
+          ref: v${{ steps.latestrelease.outputs.releasever }}
+          path: skunkworks
+      - uses: actions/checkout@v3
+        with:
+          path: action-publish-release
+      - name: Modify + Get UPDATED_PACKAGES lengths
+        id: modify-get-updated-packages
+        run: |
+          cd skunkworks
+          packages=(cli controllers)
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply .github/workflows/build-lint-test.yml:38]: Oh it's a string literal. This is a list, not a subshell.

File: .github/workflows/build-lint-test.yml [hunk 2]
```diff
@@ -16,9 +16,48 @@ jobs:
...
+          repository: MetaMask/snaps-skunkworks
+          ref: v${{ steps.latestrelease.outputs.releasever }}
+          path: skunkworks
+      - uses: actions/checkout@v3
+        with:
+          path: action-publish-release
+      - name: Modify + Get UPDATED_PACKAGES lengths
+        id: modify-get-updated-packages
+        run: |
+          cd skunkworks
```
Comment (Gudahtt) [inline .github/workflows/build-lint-test.yml:37]: Shellcheck suggests `|| exit` here in case `cd` fails

```suggestion
          cd skunkworks || exit
```

File: scripts/get-updated-packages.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+#       "version": "0.19.2"
+#     },
+#     "@metamask/snap-controllers": {
+#       "name": "@metamask/snap-controllers",
+#       "path": "packages/controllers",
+#       "version": "0.19.2"
+#     }
+#   }
+# }
+toPublish="{\"packages\":{"
```
Comment (Gudahtt) [inline scripts/get-updated-packages.sh:24]: Welp. This would have been easier to write in JavaScript/TypeScript :sweat_smile: But this works.

File: src/getReleaseNotes.ts [hunk 4]
```diff
@@ -79,33 +95,52 @@ async function getMonorepoReleaseNotes(
...
+  } else {
+    for (const [packageName, value] of Object.entries(getUpdatedPackages())) {
+      const { path: packagePath } = value;
       releaseNotes = releaseNotes.concat(
         `## ${packageName}\n\n`,
         await getPackageReleaseNotes(
           releaseVersion,
           repoUrl,
-          completeWorkspacePath,
+          String(packagePath),
```
Comment (Gudahtt) [inline src/getReleaseNotes.ts:143]: Nit: I would expect this to be a string already. If it's not a string, what is it?

```suggestion
          packagePath,
```

File: src/getReleaseNotes.test.ts [hunk 5]
```diff
@@ -40,13 +41,49 @@ jest.mock('./utils', () => {
   };
 });
 
+describe('getUpdatedPackages', () => {
+  let parseEnvVariablesMock: jest.SpyInstance;
+
+  beforeEach(() => {
+    parseEnvVariablesMock = jest.spyOn(localUtils, 'parseEnvironmentVariables');
```
Comment (Gudahtt) [inline src/getReleaseNotes.test.ts:48]: Hmm. This should already be mocked due to the line 38. It seems like you're double-mocking it here. Even without this line you should be able to use `localUtils.parseEnvironmentVariables.mock`, and to set mock implementations using `localUtils.parseEnvironmentVariables.mockImplementationOnce`, etc. At least I think so.

Comment (Gudahtt) [reply src/getReleaseNotes.test.ts:48]: I see that this pattern is pre-existing. I guess it's not a blocker then but this is not how this is supposed to work.

File: .github/workflows/build-lint-test.yml [hunk 6]
```diff
@@ -16,9 +16,48 @@ jobs:
...
+  updated-packages-test:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Get Latest Version from npm
+        id: latestrelease
+        run: echo "::set-output name=releasever::$(npm view MetaMask/snaps-skunkworks version --workspaces=false)"
+      - uses: actions/checkout@v3
+        with:
+          repository: MetaMask/snaps-skunkworks
+          ref: v${{ steps.latestrelease.outputs.releasever }}
```
Comment (mcmire) [inline .github/workflows/build-lint-test.yml:29]: Instead of targeting the latest version, we should target a specific version, so as to prevent this test from breaking if the snaps team decides to make any major changes to the project.

File: .github/workflows/build-lint-test.yml [hunk 7]
```diff
@@ -16,9 +16,48 @@ jobs:
...
+            version_increment=$(echo "$version" | awk 'BEGIN{FS=OFS="."} {$3+=1} 1')
+            jq --arg version "$version_increment" '.version = $version' "$manifest" > "$manifest_temp"
+            mv "$manifest_temp" "$manifest"
+          done
+          ../action-publish-release/scripts/get-updated-packages.sh
+      - name: Get modified UPDATED_PACKAGES lengths
+        id: get-modified-updated-packages-length
+        run: |
+          echo "::set-output name=length::$(echo '${{ steps.modify-get-updated-packages.outputs.UPDATED_PACKAGES }}' | jq '.packages | length')"
+      - name: Verify modified UPDATED_PACKAGES lengths
```
Comment (mcmire) [inline .github/workflows/build-lint-test.yml:52]: It's great that we have a test for this script, but this seems to only exercise the happy path. What if:

* the version of a package was updated but it is private
* the version of a (public) package is decremented, not just incremented
* the project is a Yarn 1.x project not just Yarn 3.x

File: src/getReleaseNotes.ts [hunk 8]
```diff
@@ -12,6 +12,21 @@ import {
...
+import { FIXED, UPDATED_PACKAGES_ERROR } from './constants';
+
+export const getUpdatedPackages = (): Record<
+  string,
+  Record<string, string>
+> => {
+  const { updatedPackages } = parseEnvironmentVariables();
+
+  if (updatedPackages === undefined) {
+    throw new Error(UPDATED_PACKAGES_ERROR);
```
Comment (mcmire) [inline src/getReleaseNotes.ts:24]: In throwing this error, we are assuming that this function is only called in the case of a monorepo with fixed versions. But because this function is extracted and lives outside the scope of `getReleaseNotes`, I think that this assumption is surprising. What do you think about moving this check to within the monorepo-with-independent-versions branch below?

File: src/getReleaseNotes.ts [hunk 9]
```diff
@@ -79,33 +95,52 @@ async function getMonorepoReleaseNotes(
...
-    const completeWorkspacePath = pathUtils.join(
-      workspaceRoot,
-      workspaceLocation,
-    );
 
-    const rawPackageManifest = await getPackageManifest(completeWorkspacePath);
-    const { name: packageName, version: packageVersion } =
-      validatePolyrepoPackageManifest(
-        rawPackageManifest,
+  if (versioningStrategy === FIXED) {
```
Comment (mcmire) [inline src/getReleaseNotes.ts:107]: The `if` statement says we are working with a monorepo (because only monorepos can have a fixed versioning scheme, or any versioning scheme, for that matter) and yet we are accessing the package manifest for a polyrepo package. I realize that you wrapped code that was already there, but to make this easier to follow, what do you think about aliasing `validatePolyrepoPackageManifest` to just `validatePackageManifest`? 

Also, since "fixed" is the default versioning strategy, what do you think about reversing these `if` statements? Check whether it's "independent" first and then fallback to "fixed".

Finally, what do you think about extracting the branches to separate functions? So you might have something like:

``` javascript
if (versioningStrategy === INDEPENDENT) {
  return await getReleaseNotesForMonorepoWithIndependentVersions();
} else {
  return await getReleaseNotesForMonorepoWithFixedVersions();
}
```

or whatever makes sense to you naming-wise.

File: src/getReleaseNotes.ts [hunk 10]
```diff
@@ -79,33 +95,52 @@ async function getMonorepoReleaseNotes(
...
+            repoUrl,
+            completeWorkspacePath,
+          ),
+          '\n\n',
+        );
+      }
+    }
+  } else {
+    for (const [packageName, value] of Object.entries(getUpdatedPackages())) {
+      const { path: packagePath } = value;
```
Comment (mcmire) [inline src/getReleaseNotes.ts:137]: You should be able to destructure this above, i.e.

``` javascript
for (const [packageName, { path: packagePath }] of Object.entries(getUpdatedPackages())) {
```

File: src/utils.ts [hunk 11]
```diff
@@ -5,21 +5,37 @@ import {
...
-  RELEASE_VERSION?: string;
+  [RELEASE_VERSION]?: string;
+  // version strategy
+  // this is set from the repository `release.config.json` key: .versioningStrategy
+  [VERSION_STRATEGY]?: string;
+  // this is a json list of the updated packages
+  [UPDATED_PACKAGES]?: string | undefined;
 }
 
+const EMPTY_PACKAGES = undefined;
```
Comment (mcmire) [inline src/utils.ts:37]: Nit: I noticed that you aren't reusing this variable in `getReleaseNotes.ts` when you check for what `updatedPackages` is, so is there a need to define this variable, i.e. why not inline `undefined` down below?

**Review by mcmire — COMMENTED** [review]
I did a first pass on this and left some comments. I can do another pass later after those are discussed.

In general, however, I feel a bit conflicted about this PR. On one hand, I appreciate that we have tried to stick to the existing code and not rock the boat too much. There aren't too many changes here and that's good. On the other hand, I was really hoping that we could move the existing code to TypeScript. Between the steps in `action.yml`, the shell scripts in `scripts/`, and the JavaScript code, I find it a bit challenging to follow the flow of this action — what sort of situations it's designed to cover and what it does in those situations. There were a couple spots in the shell code that I had to re-read a few times to fully grok. Also, while it's great that we have tests for sections of this action, they are limited in scope. There's nothing that really tests the action all the way through, and I feel like we ought to have that. Such a test, or a collection of tests, would also help someone who is new to this whole project, because it would remove the magic — it would give them a clear picture of how the action works, from the very top, and how a project designed to work with this action would look like. Of course in order to prevent the GitHub release from actually getting created, we would have to stub the command that gets run at the end, and that would require moving everything to TypeScript, so that's where that comes into play.

Maybe I could have helped you more with this, or maybe you only had so much time, and that's fine. Without these tests, though, I think we will just have to be more meticulous in how we update this action, and we will have to test it manually in various scenarios to ensure that we don't discover bugs when attempting to publish a new release later on. It wouldn't be the end of the world if that happened, of course, but fixing bugs with "devops-y stuff" on the spot is not the most enjoyable thing.

In summary, in case it gets misunderstood, I'm not rejecting anything here. I will do another pass on this later to verify that the flows here make sense. In the future perhaps we can do a refactor using some of the tools I've developed in `create-release-branch` but in the meantime I think we just have to be a bit careful.

File: src/getReleaseNotes.test.ts [hunk 12]
```diff
@@ -221,6 +269,66 @@ describe('getReleaseNotes', () => {
...
+  it('should get the release notes for monorepos (independent)', async () => {
+    const mockWorkspaceRoot = 'foo/';
+    const mockRepoUrl = 'https://github.com/Org/Name';
+    const mockVersion = '1.0.0';
+    const mockWorkspaces = ['a', 'b', 'c'];
+    const mockChangelog = 'a changelog';
+    const mockVersionStrategy = 'independent';
+
+    getUpdatedPackagesMock.mockImplementationOnce(() => {
+      return JSON.parse(MOCK_UPDATED_PACAKGES);
```
Comment (Gudahtt) [inline src/getReleaseNotes.test.ts:281]: This mock looks wrong. This function doesn't return this object directly (which is of the form `packages: { [name]: [manifest], ...}`. It's returning the `packages` value directly, i.e. `{ [name]: [manifest], ...}``

File: .github/workflows/build-lint-test.yml (hunk 6 above)
Comment (rickycodes) [reply .github/workflows/build-lint-test.yml:29]: this test will only work against the latest published version. if we pin it to an older version it will fail because the next version will have already been published.

Comment (rickycodes) [reply .github/workflows/build-lint-test.yml:29]: this has been altered now to get values using `yarn workspaces list ...` and pick the first two (which still ends up being "cli" and "controllers", but now those packages/names are free to change without breaking this)

Comment (mcmire) [reply .github/workflows/build-lint-test.yml:29]: Nice! That looks a lot better.

File: .github/workflows/build-lint-test.yml (hunk 7 above)
Comment (rickycodes) [reply .github/workflows/build-lint-test.yml:72]: > the version of a package was updated but it is private

I've now accounted for this as well.

Comment (mcmire) [reply .github/workflows/build-lint-test.yml:83]: If I recall correctly, we decided to drop Yarn 1.x support so that test shouldn't be necessary.

File: .github/workflows/build-lint-test.yml [hunk 13]
```diff
@@ -16,9 +16,76 @@ jobs:
...
+                if [[ "$private" == true ]]; then
+                  packages+=("$location")
+                  count=$((count + 1))
+                fi
+              fi
+            fi
+            if [[ "$count" -eq 3 ]]; then
+              break;
+            fi
+          done< <(echo "$workspaces" | jq --raw-output '"\(.location) \(.name)"')
```
Comment (mcmire) [inline .github/workflows/build-lint-test.yml:59]: We seem to be doing the same thing for public packages as for private packages:

``` shell
                if [[ "$private" != true ]]; then
                  packages+=("$location")
                  count=$((count + 1))
                fi
                if [[ "$private" == true ]]; then
                  packages+=("$location")
                  count=$((count + 1))
                fi
```

Maybe we want to have two loops, one that goes over just the public packages, one that goes over just the private packages? Like:

``` suggestion
          while read -r location name private; do
            if [[ $name != "root" && $private != true && $count -le 2 ]]; then
              packages+=("$location")
              count=$((count + 1))
            fi
          done< <(echo "$workspaces" | jq --raw-output '"\(.location) \(.name) \(.private)"')

          while read -r location name private; do
            if [[ $name != "root" && $private == true && $count -le 3 ]]; then
              packages+=("$location")
              count=$((count + 1))
            fi
          done< <(echo "$workspaces" | jq --raw-output '"\(.location) \(.name) \(.private)"')
```

File: scripts/create-github-release.sh [hunk 14]
```diff
@@ -9,14 +9,35 @@ if [[ -z $RELEASE_NOTES ]]; then
...
+fi
+
 gh release create \
   "v$RELEASE_VERSION" \
   --title "$RELEASE_VERSION" \
   --notes "$RELEASE_NOTES"
+
+if [[ "$(jq 'has("workspaces")' package.json)" = "true" && "$VERSION_STRATEGY" = "independent"  ]]; then
+
+  if [[ -z $UPDATED_PACKAGES ]]; then
```
Comment (mcmire) [inline scripts/create-github-release.sh:29]: Hmm actually, what do you think about making this check up front, before we create a GitHub release above? Otherwise we could leave things in a mixed state.

File: src/utils.test.ts [hunk 15]
```diff
@@ -62,13 +63,26 @@ describe('parseEnvironmentVariables', () => {
...
       parseEnvironmentVariables({
         GITHUB_WORKSPACE: 'foo',
         REPOSITORY_URL: 'https://github.com/MetaMask/snaps-skunkworks.git',
         RELEASE_VERSION: 'kaplar',
       }),
-    ).toThrow(RELEASE_VERSION_ERROR);
+    ).toThrow('process.env.RELEASE_VERSION must be a valid SemVer version.');
+  });
+
+  it('throws if VERSION_STRATEGY is invalid', () => {
```
Comment (mcmire) [inline src/utils.test.ts:76]: s/VERSION_STRATEGY/RELEASE_STRATEGY ?

File: src/utils.ts [hunk 16]
```diff
@@ -18,6 +14,11 @@ interface ExpectedProcessEnv extends Partial<Record<string, string>> {
   // The version to be released,
   // this is set from the repository `package.json` key: .version
   RELEASE_VERSION?: string;
+  // version strategy
+  // this is set from the repository `release.config.json` key: .versioningStrategy
+  VERSION_STRATEGY?: string;
+  // this is a json list of the updated packages
+  UPDATED_PACKAGES?: string | undefined;
```
Comment (mcmire) [inline src/utils.ts:21]: Isn't the `| undefined` covered by the `?`?

File: src/utils.test.ts (hunk 15 above)
Comment (rickycodes) [reply src/utils.test.ts:76]: yeah? maybe "RELEASE_" all the things makes more sense. I'll update

File: src/utils.ts (hunk 16 above)
Comment (rickycodes) [reply src/utils.ts:21]: ah good catch. you're right that's redundant

File: scripts/create-github-release.sh (hunk 14 above)
Comment (rickycodes) [reply scripts/create-github-release.sh]: ah right. moved it back!

File: .github/workflows/build-lint-test.yml (hunk 13 above)
Comment (rickycodes) [reply .github/workflows/build-lint-test.yml]: I've updated this now and I think what's there is a bit easier to digest now.

File: .github/workflows/build-lint-test.yml (hunk 7 above)
Comment (rickycodes) [reply .github/workflows/build-lint-test.yml]: this is included now

File: scripts/get-updated-packages.sh [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+# store initial length of toPublish
+len="${#toPublish}"
+
+workspaces=$(yarn workspaces list --verbose --json)
```
[the remark this answers is not in the document]
Comment (mcmire) [reply scripts/get-updated-packages.sh]: I believe we decided that projects which want to use the new version of this action need to use Yarn 3.x. Does that sound correct?

File: .github/workflows/build-lint-test.yml [hunk 18]
```diff
@@ -16,9 +16,85 @@ jobs:
...
+            VERSION=$(jq --raw-output .version "$MANIFEST")
+            NEW_VERSION=$(echo "$VERSION" | awk 'BEGIN{FS=OFS="."} {$3+=1} 1')
+            if [[ "$2" == "unbump" ]]; then
+              IFS='.' read -r -a VERSIONS <<< "$VERSION"
+              if [[ "${VERSIONS[2]}" == "0" ]]; then
+                NEW_VERSION=$(echo "$VERSION" | awk 'BEGIN{FS=OFS="."} {$2-=1} 1' | awk 'BEGIN{FS=OFS="."} {$3=9} 1')
+              else
+                NEW_VERSION=$(echo "$VERSION" | awk 'BEGIN{FS=OFS="."} {$3-=1} 1')
+              fi
+            fi;
```
Comment (mcmire) [inline .github/workflows/build-lint-test.yml:72]: It looks like line 70 will negate line 64, so the version won't be changed at all. Is that intentional?

Additionally, I'm not understanding the "1" at the end of the awk script. Curiously, removing this resulting in nothing being printed, but I'm not sure what it does.

I am wondering whether this would be more straightforward using bash rather than awk... What do you think about this?

``` suggestion
            IFS='.' read -r -a VERSIONS <<< "$VERSION"
            MAJOR="${VERSIONS[0]}"
            MINOR="${VERSIONS[1]}"
            PATCH="${VERSIONS[2]}"
            if [[ "$2" == "unbump" ]]; then
              if [[ "$PATCH" == "0" ]]; then
                if [[ "$MINOR" == "0" ]]; then
                  # e.g. 10.0.0 -> 9.9.9
                  NEW_VERSION="$((MAJOR - 1)).9.9"
                else
                  # e.g. 1.10.0 -> 1.9.9
                  NEW_VERSION="${MAJOR}.$((MINOR - 1)).9"
                fi
              else
                # e.g. 1.2.10 -> 1.2.9
                NEW_VERSION="${MAJOR}.${MINOR}.$((PATCH - 1))"
              fi
            else
              # e.g. 1.0.0 -> 1.0.1
              NEW_VERSION="${MAJOR}.${MINOR}.$((PATCH + 1))"
            fi
```

File: src/getReleaseNotes.test.ts [hunk 19]
```diff
@@ -221,6 +266,69 @@ describe('getReleaseNotes', () => {
...
+    readFileMock.mockImplementation(
+      async (path: string) =>
+        `${mockChangelog} for ${path.charAt(
+          path.indexOf('/CHANGELOG.md') - 1,
+        )}`,
+    );
+
+    parseChangelogMock.mockImplementation(parseChangelogMockImplementation);
+
+    await releaseNotesUtils.getReleaseNotes();
```
[the remark this answers is not in the document]
Comment (mcmire) [reply src/getReleaseNotes.test.ts:327]: Okay, these items are both addressed. I think it would make the tests clearer and easier to update later if, for the `getReleaseNotes` tests, we had `describe`s for polyrepo, monorepo w/ independent, and monorepo w/ fixed cases, and then we tested both the happy path and the sad path ("The computed release notes are empty") for each of these setups, but, I guess it's not critical.

**Review by mcmire — APPROVED** [review]
LGTM.

File: src/getReleaseNotes.ts (hunk 8 above)
Comment (mcmire) [reply src/getReleaseNotes.ts]: Hmm, I guess there has to be a check here because `updatedPackages` can be undefined. Alright.

**Comment by mcmire** [comment]
I've been testing this out in three cases:

* Monorepo with fixed versions, using `snaps-monorepo`
* Monorepo with independent versions, using `controllers` (`monorepo-poc` branch). *Note that we are no longer using `action-create-release-pr` to create release PRs but `create-release-branch` instead.*
* Polyrepo package

I encountered a problem when attempting to release `snaps-monorepo`: https://github.com/mcmire/snaps-monorepo/actions/runs/3252758730/jobs/5339319338. I made fixes in 2cc1e0a...7e70bcd. Here is the successful release: https://github.com/mcmire/snaps-monorepo/actions/runs/3253016948/jobs/5339866555

I also encountered a problem with `controllers`: https://github.com/mcmire/controllers/actions/runs/3253182761/jobs/5340203034. I made fixes in 7e70bcd...021618a. That worked here: https://github.com/mcmire/controllers/actions/runs/3253390324/jobs/5340619148. However, I realized that `controllers` was not using the latest version of `npm-publish` (which adds support for independent monorepos). Upon attempting to run `create-release-branch` again, however, I ran into this bug: https://github.com/MetaMask/create-release-branch/issues/43. We need to fix this before we can continue testing this action.

I haven't gotten around to testing the polyrepo package case yet.

**Comment by rickycodes** [comment]
I also tested this on a version of `snaps-monorepo` where I added `release.config.json` with an `"independent"` `.versioningStrategy` and that appears to have run correctly:

- It correctly ran the newly added `get-release-packages.sh` [here](https://github.com/rickycodes/snaps-monorepo/actions/runs/3379184167/jobs/5610407904#step:5:24)
- It correctly determined the `RELEASE_STRATEGY` as `independent` [here](https://github.com/rickycodes/snaps-monorepo/actions/runs/3379184167/jobs/5610407904#step:5:264)
- It correctly got a complete list of the packages to be published (in this case all since I did this with `Create Release Pull Request`) [here](https://github.com/rickycodes/snaps-monorepo/actions/runs/3379184167/jobs/5610407904#step:5:267)
- It correctly created a single version tag and individual tags for each package [here](https://github.com/rickycodes/snaps-monorepo/tags)

**Review by Gudahtt — APPROVED** [review]
LGTM!
